node/src/store/header_ranges.rs

```diff
@@ -0,0 +1,455 @@
+        if *to_insert.start() > head_range.end() + 1 {
+            return Ok(RangeScanResult {
+                range_index: self.0.len(),
+                range: to_insert.clone(),
+                range_to_remove: None,
+            });
+        }
+
+        let mut stored_ranges_iter = self.0.iter().enumerate();
+        let mut found_range = loop {
+            let Some((idx, stored_range)) = stored_ranges_iter.next() else {
+                return Err(StoreError::InsertPlacementDisallowed(
+                    *to_insert.start(),
+                    *to_insert.end(),
+                ));
+            };
+
+            // TODO: break early
+            //if sth
+
+            if let Some(intersection) = ranges_intersection(stored_range, to_insert) {
+                return Err(StoreError::HeaderRangeOverlap(
+                    *intersection.start(),
+                    *intersection.end(),
+                ));
+            }
+
+            if let Some(consolidated) = try_consolidate_ranges(stored_range, to_insert) {
+                break RangeScanResult {
+                    range_index: idx,
+                    range: consolidated,
+                    range_to_remove: None,
+                };
+            }
+        };
+
+        // we have a hit, check whether we can merge with the next range too
+        if let Some((idx, range_after)) = stored_ranges_iter.next() {
+            if let Some(intersection) = ranges_intersection(range_after, to_insert) {
+                return Err(StoreError::HeaderRangeOverlap(
+                    *intersection.start(),
+                    *intersection.end(),
+                ));
+            }
+
+            if let Some(consolidated) = try_consolidate_ranges(range_after, &found_range.range) {
+                found_range = RangeScanResult {
+                    range_index: found_range.range_index,
+                    range: consolidated,
+                    range_to_remove: Some(idx),
+                };
+            }
+        }
+
+        Ok(found_range)
+    }
+
+    /// Return whether range is empty
+    fn is_empty(&self) -> bool {
+        self.0.iter().all(|r| r.is_empty())
+    }
+
+    /// Return highest height in the range
+    fn head(&self) -> Option<u64> {
+        self.0.last().map(|r| *r.end())
+    }
+}
+
+impl HeaderRanges {
+    /// Return whether `headerRanges` contains provided height
+    pub fn contains(&self, height: u64) -> bool {
+        self.0.iter().any(|r| r.contains(&height))
+    }
+}
+
+impl Display for HeaderRanges {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "[")?;
+        for (idx, range) in self.0.iter().enumerate() {
+            if idx == 0 {
+                write!(f, "{}..={}", range.start(), range.end())?;
+            } else {
+                write!(f, ", {}..={}", range.start(), range.end())?;
+            }
+        }
+        write!(f, "]")
```

Comment:
I prefer something like `[10-11, 14-20]`.

Reply:
Changed from rust-like notation, to more human recognisable `[1-2, 10-20]`